File: atividade3_javascript/Block.js
'use strict'

class Block {
    constructor(data, hash_anterior, nonce) {
        this.data = data;
        this.hash = "";
        this.hash_anterior = hash_anterior;
        this.timestamp = new Date().toUTCString();
        this.nonce = nonce;
    }
    // cria um nó zero para a blockchain
    zero() {
        return new Block("bloco zero", "primeiro hash", 0);
    }
    // cria um bloco com valores informados
    new(data, hash_anterior, nonce) {
        return new Block(data, hash_anterior, nonce)
    }

    validar() {}
}

module.exports = Block;

File: atividade3/lib/Block.ex
defmodule Atividade3.Block do
  # struct do bloco
  defstruct [:data, :timestamp, :hash_anterior, :hash]

  # cria o primeiro bloco da cadeia
  def zero do
    %Atividade3.Block {
      data: "primeiro",
      hash_anterior: "primeiro_hash_da_cadeia",
      timestamp: NaiveDateTime.utc_now
    }
  end

  # constroi novo bloco a partir dos dados
  # e do hash anterior
  def new(data, hash_anterior) do
    %Atividade3.Block{
      data: data,
      hash_anterior: hash_anterior,
      timestamp: NaiveDateTime.utc_now,
    }
  end

  # checa se o bloco é valido
    def valido?(%Atividade3.Block{} = block) do
      Atividade3.Crypto.hash(block) == block.hash
    end

    def valido?(
      %Atividade3.Block{} = block,
      %Atividade3.Block{} = prev_block
    ) do
      (block.hash_anterior == prev_block.hash) && valido?(block)
    end
end
